test_callbacks_non_minus_one_return_continues_transfer fails intermittently on macOS

When building wheels on macOS-14, the `test_callbacks_non_minus_one_return_continues_transfer` is failing consistently:

```
  =================================== FAILURES ===================================
  ___ test_callbacks_non_minus_one_return_continues_transfer[timer_returns_42] ___
  
  install_callbacks = <function install_callbacks.<locals>.make at 0x104bdab60>
  socket_cb = <function _ok at 0x1036f22a0>
  timer_cb = <function <lambda> at 0x1036f2520>
  
      @pytest.mark.parametrize(
          "socket_cb,timer_cb",
          [
              (_none, _none),
              (_ok, _ok),
              (lambda *_: 42, _ok),
              (_ok, lambda _: 42),
          ],
          ids=["return_None", "return_0", "socket_returns_42", "timer_returns_42"],
      )
      def test_callbacks_non_minus_one_return_continues_transfer(
          install_callbacks, socket_cb, timer_cb
      ):
          easy, multi = install_callbacks(socket_cb, timer_cb)
          ok, err = _drive_to_completion(easy, multi)
  >       assert ok and not err
  E       assert (None)
  
  /Users/runner/work/pycurl/pycurl/tests/multi_callback_test.py:293: AssertionError
```
@aeroyorch do you mind taking a look?

--- a/tests/multi_callback_test.py
+++ b/tests/multi_callback_test.py
@@ -7,10 +7,9 @@ from urllib.parse import urlencode
 import logging
 import pycurl
 import pytest
-import select
-import sys
 import time
 from . import util
+from .multi_driver import TimerState, install_timer_tracker, pump
 
 logger = logging.getLogger(__name__)
 
@@ -25,37 +24,14 @@ class MultiCtx:
     sockets: dict[int, int] = field(default_factory=dict)
 
     handle_added: bool = False
-    timer_pending: bool = False
+    timer_state: TimerState = field(default_factory=TimerState)
 
     write_calls: int = 0
     bytes_received: int = 0
 
 
 def _event_loop_step(ctx: MultiCtx, timeout: float = 0.2) -> None:
-    ctx.multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
-
-    rset, wset, xset = ctx.multi.fdset()
-
-    if not (rset or wset or xset):
-        time.sleep(min(timeout, 0.01))
-        return
-
-    r_ready, w_ready, x_ready = select.select(rset, wset, xset, timeout)
-
-    actions: dict[int, int] = {}
-    for s in r_ready:
-        actions[s] = actions.get(s, 0) | pycurl.CSELECT_IN
-    for s in w_ready:
-        actions[s] = actions.get(s, 0) | pycurl.CSELECT_OUT
-    for s in x_ready:
-        actions[s] = actions.get(s, 0) | pycurl.CSELECT_ERR
-
-    for s, act in actions.items():
-        ctx.multi.socket_action(s, act)
-
-    if ctx.timer_pending:
-        ctx.timer_pending = False
-        ctx.multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
+    pump(ctx.multi, ctx.timer_state, timeout)
 
 
 def _run_until(
@@ -125,11 +101,9 @@ def multi_ctx(app) -> Generator[MultiCtx, None, None]:
     def timer_callback(timeout_ms: int) -> None:
         logger.debug("timer_callback: timeout=%d", timeout_ms)
         ctx.timer_result = timeout_ms
-        if timeout_ms == 0:
-            ctx.timer_pending = True
 
     multi.setopt(pycurl.M_SOCKETFUNCTION, socket_callback)
-    multi.setopt(pycurl.M_TIMERFUNCTION, timer_callback)
+    ctx.timer_state = install_timer_tracker(multi, timer_callback)
 
     try:
         yield ctx
@@ -214,7 +188,6 @@ def test_easy_pause_unpause(multi_ctx: MultiCtx, app):
 
 @pytest.fixture
 def install_callbacks(app):
-    """Yield make(socket_cb, timer_cb) -> (easy, multi); cleans up on exit."""
     created: list[tuple[pycurl.Curl, pycurl.CurlMulti]] = []
 
     def make(socket_cb, timer_cb):
@@ -223,9 +196,9 @@ def install_callbacks(app):
         easy.setopt(pycurl.WRITEFUNCTION, BytesIO().write)
         multi = pycurl.CurlMulti()
         multi.setopt(pycurl.M_SOCKETFUNCTION, socket_cb)
-        multi.setopt(pycurl.M_TIMERFUNCTION, timer_cb)
+        timer_state = install_timer_tracker(multi, timer_cb)
         created.append((easy, multi))
-        return easy, multi
+        return easy, multi, timer_state
 
     yield make
 
@@ -238,29 +211,15 @@ def install_callbacks(app):
         easy.close()
 
 
-def _drive_to_completion(easy, multi, timeout=5.0):
+def _drive_to_completion(easy, multi, timer_state, timeout=5.0):
     multi.add_handle(easy)
-    multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
     deadline = time.monotonic() + timeout
     while time.monotonic() < deadline:
+        pump(multi, timer_state)
         _, ok_list, err_list = multi.info_read()
         if ok_list or err_list:
             return ok_list, err_list
-        rset, wset, xset = multi.fdset()
-        if not (rset or wset or xset):
-            time.sleep(0.01)
-            continue
-        r, w, x = select.select(rset, wset, xset, 0.2)
-        actions: dict[int, int] = {}
-        for s in r:
-            actions[s] = actions.get(s, 0) | pycurl.CSELECT_IN
-        for s in w:
-            actions[s] = actions.get(s, 0) | pycurl.CSELECT_OUT
-        for s in x:
-            actions[s] = actions.get(s, 0) | pycurl.CSELECT_ERR
-        for s, act in actions.items():
-            multi.socket_action(s, act)
-    return None, None
+    pytest.fail(f"transfer did not complete within {timeout}s")
 
 
 def _ok(*_args, **_kwargs):
@@ -276,7 +235,6 @@ def _raises_runtime(*_args, **_kwargs):
 
 
 # Only -1 aborts; any other return (including non-zero ints) continues.
-@pytest.mark.skipif(sys.platform == "darwin", reason="https://github.com/pycurl/pycurl/issues/984")
 @pytest.mark.parametrize(
     "socket_cb,timer_cb",
     [
@@ -290,8 +248,8 @@ def _raises_runtime(*_args, **_kwargs):
 def test_callbacks_non_minus_one_return_continues_transfer(
     install_callbacks, socket_cb, timer_cb
 ):
-    easy, multi = install_callbacks(socket_cb, timer_cb)
-    ok, err = _drive_to_completion(easy, multi)
+    easy, multi, timer_state = install_callbacks(socket_cb, timer_cb)
+    ok, err = _drive_to_completion(easy, multi, timer_state)
     assert ok and not err
 
 
@@ -315,7 +273,7 @@ def test_callbacks_non_minus_one_return_continues_transfer(
 def test_callbacks_failure_aborts_transfer(
     install_callbacks, capfd, socket_cb, timer_cb, expected_stderr
 ):
-    easy, multi = install_callbacks(socket_cb, timer_cb)
+    easy, multi, _ = install_callbacks(socket_cb, timer_cb)
     with pytest.raises(pycurl.error):
         multi.add_handle(easy)
         multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)

--- a/tests/multi_driver.py
+++ b/tests/multi_driver.py
@@ -1,0 +1,65 @@
+"""Helpers for driving curl_multi_socket_action in tests."""
+
+from __future__ import annotations
+
+import select
+import time
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
+
+import pycurl
+
+
+@dataclass
+class TimerState:
+    pending: bool = False
+
+
+def install_timer_tracker(
+    multi: pycurl.CurlMulti,
+    base_cb: Callable[[int], Any] | None = None,
+) -> TimerState:
+    """Install M_TIMERFUNCTION; record timeout_ms == 0 in a TimerState.
+
+    `base_cb`, if provided, is called and its return value forwarded to
+    libcurl, so callers can install an arbitrary user callback without
+    losing the pending-timer signal.
+    """
+    state = TimerState()
+
+    def timer(timeout_ms: int) -> Any:
+        if timeout_ms == 0:
+            state.pending = True
+        return None if base_cb is None else base_cb(timeout_ms)
+
+    multi.setopt(pycurl.M_TIMERFUNCTION, timer)
+    return state
+
+
+def pump(multi: pycurl.CurlMulti, timer_state: TimerState, timeout: float = 0.2) -> int:
+    """Run one event-loop tick; return the number of running handles."""
+    _, running = multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
+    rset, wset, xset = multi.fdset()
+    if not (rset or wset or xset):
+        if timer_state.pending:
+            timer_state.pending = False
+            _, running = multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
+        time.sleep(min(timeout, 0.01))
+        return running
+
+    r, w, x = select.select(rset, wset, xset, timeout)
+    actions = {}
+    for s in r:
+        actions[s] = actions.get(s, 0) | pycurl.CSELECT_IN
+    for s in w:
+        actions[s] = actions.get(s, 0) | pycurl.CSELECT_OUT
+    for s in x:
+        actions[s] = actions.get(s, 0) | pycurl.CSELECT_ERR
+    for s, act in actions.items():
+        _, running = multi.socket_action(s, act)
+
+    if timer_state.pending:
+        timer_state.pending = False
+        _, running = multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
+    return running

--- a/tests/multi_socket_test.py
+++ b/tests/multi_socket_test.py
@@ -1,6 +1,5 @@
 import gc
 import logging
-import select
 import sys
 import time
 import weakref
@@ -11,6 +10,7 @@ import pycurl
 import pytest
 
 from . import util
+from .multi_driver import install_timer_tracker, pump
 
 logger = logging.getLogger(__name__)
 
@@ -24,53 +24,13 @@ def multi():
         m.close()
 
 
-def _setup_timer(multi):
-    timer_state = {"pending": False}
-
-    def timer(timeout_ms):
-        if timeout_ms == 0:
-            timer_state["pending"] = True
-
-    multi.setopt(pycurl.M_TIMERFUNCTION, timer)
-    return timer_state
-
-
-def _consume_timer(multi, timer_state):
-    if timer_state and timer_state["pending"]:
-        timer_state["pending"] = False
-        multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
-
-
-def _drive_multi(multi, timeout=0.2, timer_state=None):
-    _, running = multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
-
-    rset, wset, xset = multi.fdset()
-    if not (rset or wset or xset):
-        _consume_timer(multi, timer_state)
-        time.sleep(min(timeout, 0.01))
-        return running
-
-    r_ready, w_ready, x_ready = select.select(rset, wset, xset, timeout)
-    actions = {}
-    for s in r_ready:
-        actions[s] = actions.get(s, 0) | pycurl.CSELECT_IN
-    for s in w_ready:
-        actions[s] = actions.get(s, 0) | pycurl.CSELECT_OUT
-    for s in x_ready:
-        actions[s] = actions.get(s, 0) | pycurl.CSELECT_ERR
-
-    for s, act in actions.items():
-        _, running = multi.socket_action(s, act)
-
-    _consume_timer(multi, timer_state)
-    return running
-
-
 def _find_socket(multi, timeout=5.0, timer_state=None):
     deadline = time.monotonic() + timeout
     while time.monotonic() < deadline:
         multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
-        _consume_timer(multi, timer_state)
+        if timer_state and timer_state.pending:
+            timer_state.pending = False
+            multi.socket_action(pycurl.SOCKET_TIMEOUT, 0)
         rset, wset, xset = multi.fdset()
         if rset or wset or xset:
             return (rset or wset or xset)[0]
@@ -97,7 +57,7 @@ def _chunks_transfer(app, multi, socket_callback, label):
     per-iteration work (e.g. assign() from outside the callback). The easy
     handle is removed and closed on exit.
     """
-    timer_state = _setup_timer(multi)
+    timer_state = install_timer_tracker(multi)
     multi.setopt(pycurl.M_SOCKETFUNCTION, socket_callback)
 
     c = util.DefaultCurl()
@@ -111,7 +71,7 @@ def _chunks_transfer(app, multi, socket_callback, label):
         deadline = time.monotonic() + 10.0
         while running:
             _assert_within_deadline(deadline, label)
-            running = _drive_multi(multi, timeout=0.2, timer_state=timer_state)
+            running = pump(multi, timer_state, timeout=0.2)
             yield timer_state
             multi.select(0.1)
     finally:
@@ -130,7 +90,7 @@ def test_multi_socket(app, multi):
     ]
 
     socket_events = []
-    timer_state = _setup_timer(multi)
+    timer_state = install_timer_tracker(multi)
 
     # socket callback
     def socket(event, socket, multi_handle, data):
@@ -157,7 +117,7 @@ def test_multi_socket(app, multi):
     deadline = time.monotonic() + 10.0
     while running:
         _assert_within_deadline(deadline, "multi socket transfer")
-        running = _drive_multi(multi, timeout=0.2, timer_state=timer_state)
+        running = pump(multi, timer_state, timeout=0.2)
         # currently no more I/O is pending, could do something in the meantime
         # (display a progress bar, etc.)
         multi.select(0.1)
@@ -335,7 +295,7 @@ def test_clear_assignment_inside_socket_callback_releases_ref(app, multi, clear)
 
 def _assign_marker_then_close(app):
     multi = pycurl.CurlMulti()
-    timer_state = _setup_timer(multi)
+    timer_state = install_timer_tracker(multi)
 
     class Marker:
         pass
@@ -359,7 +319,7 @@ def _assign_marker_then_close(app):
 
     deadline = time.monotonic() + 5.0
     while time.monotonic() < deadline and not state["assigned"]:
-        _drive_multi(multi, timeout=0.1, timer_state=timer_state)
+        pump(multi, timer_state, timeout=0.1)
 
     assert state["assigned"], "did not assign marker before timeout"
 

